File: src/db-toolkit/operations/schema_explorer.py
"""Schema exploration operations."""

from typing import Dict, List, Any, Optional
from ..connectors.factory import ConnectorFactory
from ..core.models import DatabaseConnection
from ..utils.cache import SchemaCache


class SchemaExplorer:
    """Handles database schema exploration and metadata fetching."""
    
    def __init__(self):
        """Initialize schema explorer."""
        self.cache = SchemaCache()
    
    async def get_schema_tree(self, connection: DatabaseConnection, use_cache: bool = True) -> Dict[str, Any]:
        """Get complete schema tree for connection."""
        cache_key = f"{connection.id}_schema"
        
        if use_cache:
            cached = self.cache.get(cache_key)
            if cached:
                return cached
        
        try:
            connector = ConnectorFactory.create_connector(connection.db_type)
            await connector.connect(connection)
            
            schema_tree = {
                "connection_id": connection.id,
                "db_type": connection.db_type.value,
                "schemas": {}
            }
            
            schemas = await connector.get_schemas()
            
            for schema_name in schemas:
                tables = await connector.get_tables(schema_name)
                schema_tree["schemas"][schema_name] = {
                    "tables": {},
                    "table_count": len(tables)
                }
                
                for table_name in tables:
                    columns = await connector.get_columns(table_name, schema_name)
                    schema_tree["schemas"][schema_name]["tables"][table_name] = {
                        "columns": columns,
                        "column_count": len(columns)
                    }
            
            await connector.disconnect()
            
            # Cache the result
            self.cache.set(cache_key, schema_tree)
            
            return schema_tree
            
        except Exception as e:
            return {"error": str(e), "success": False}
    
    async def get_table_info(self, connection: DatabaseConnection, schema: str, table: str) -> Dict[str, Any]:
        """Get detailed table information."""
        try:
            connector = ConnectorFactory.create_connector(connection.db_type)
            await connector.connect(connection)
            
            columns = await connector.get_columns(table, schema)
            
            # Get sample data (first 5 rows)
            sample_query = self._build_sample_query(connection.db_type, schema, table)
            sample_result = await connector.execute_query(sample_query)
            
            await connector.disconnect()
            
            return {
                "success": True,
                "table": table,
                "schema": schema,
                "columns": columns,
                "sample_data": sample_result.get("data", [])[:5] if sample_result.get("success") else []
            }
            
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    def _build_sample_query(self, db_type: str, schema: str, table: str) -> str:
        """Build sample data query for different database types."""
        if db_type == "mongodb":
            return "{}"  # Empty filter for MongoDB
        elif db_type == "sqlite":
            return f"SELECT * FROM {table} LIMIT 5"
        else:
            return f"SELECT * FROM {schema}.{table} LIMIT 5"
    
    async def refresh_schema(self, connection_id: str):
        """Refresh cached schema for connection."""
        cache_key = f"{connection_id}_schema"
        self.cache.delete(cache_key)
    
    def get_cached_schemas(self) -> List[str]:
        """Get list of cached schema keys."""
        return self.cache.get_keys()

File: src/db-toolkit-ui/electron/backend/operations/schema-explorer.js
/**
 * Schema exploration operations.
 */

const { connectionManager } = require('../utils/connection-manager');
const { schemaCache } = require('../utils/cache');
const { CACHE_TTL } = require('../utils/constants');
const { logger } = require('../utils/logger.js');

class SchemaExplorer {
  async getSchemaTree(connection, useCache = true) {
    const cacheKey = `${connection.id}_schema`;

    logger.info(`Getting schema tree for connection '${connection.name}' (use_cache=${useCache})`);

    if (useCache) {
      const cached = schemaCache.get(cacheKey);
      if (cached) {
        logger.info(`Returning cached schema for '${connection.name}'`);
        return cached;
      }
    }

    try {
      let connector = await connectionManager.getConnector(connection.id);
      if (!connector) {
        const success = await connectionManager.connect(connection);
        if (!success) {
          throw new Error('Failed to establish database connection');
        }
        connector = await connectionManager.getConnector(connection.id);
        if (!connector) {
          throw new Error('Connection manager failed to provide connector');
        }
      }

      const schemaTree = {
        connection_id: connection.id,
        db_type: connection.db_type,
        schemas: {},
      };

      const schemas = await connector.getSchemas();
      logger.info(`Found ${schemas.length} schemas:`, schemas);

      for (const schemaName of schemas) {
        const tables = await connector.getTables(schemaName);
        logger.info(`Schema '${schemaName}' has ${tables.length} tables:`, tables);
        schemaTree.schemas[schemaName] = {
          tables: {},
          table_count: tables.length,
        };

        for (const tableName of tables) {
          const columns = await connector.getColumns(tableName, schemaName);
          schemaTree.schemas[schemaName].tables[tableName] = {
            columns,
            column_count: columns.length,
          };
        }
      }

      schemaCache.set(cacheKey, schemaTree, CACHE_TTL.SCHEMA_TREE);
      return schemaTree;
    } catch (error) {
      logger.error(`Failed to get schema tree for '${connection.name}':`, error);
      return { error: error.message, success: false };
    }
  }

  async getTableInfo(connection, schema, table) {
    const cacheKey = `${connection.id}_table_${schema}_${table}`;
    const cached = schemaCache.get(cacheKey);
    if (cached) return cached;

    try {
      let connector = await connectionManager.getConnector(connection.id);
      if (!connector) {
        const success = await connectionManager.connect(connection);
        if (!success) {
          throw new Error('Failed to establish database connection');
        }
        connector = await connectionManager.getConnector(connection.id);
        if (!connector) {
          throw new Error('Connection manager failed to provide connector');
        }
      }

      const columns = await connector.getColumns(table, schema);

      const sampleQuery = this.buildSampleQuery(connection.db_type, schema, table);
      const sampleResult = await connector.executeQuery(sampleQuery);

      const tableInfo = {
        success: true,
        table,
        schema,
        columns,
        sample_data: sampleResult.success ? (sampleResult.data || []).slice(0, 5) : [],
      };

      schemaCache.set(cacheKey, tableInfo, CACHE_TTL.TABLE_INFO);
      return tableInfo;
    } catch (error) {
      logger.error(`Failed to get table info for '${connection.name}.${schema}.${table}':`, error);
      return { success: false, error: error.message };
    }
  }

  buildSampleQuery(dbType, schema, table) {
    if (dbType === 'mongodb') {
      return '{}';
    } else if (dbType === 'sqlite') {
      return `SELECT * FROM "${table}" LIMIT 5`;
    } else {
      return `SELECT * FROM ${schema}."${table}" LIMIT 5`;
    }
  }

  async refreshSchema(connectionId) {
    const keysToRemove = schemaCache.getKeys().filter(key => key.startsWith(`${connectionId}_`));
    keysToRemove.forEach(key => schemaCache.delete(key));
  }

  getCachedSchemas() {
    return schemaCache.getKeys();
  }
}

const schemaExplorer = new SchemaExplorer();

module.exports = { SchemaExplorer, schemaExplorer, schemaCache };
